backend/oauth.go

```diff
@@ -0,0 +1,236 @@
+		return
+	}
+
+	code := query.Get("code")
+	if code == "" {
+		http.Error(w, "Missing OAuth2 code", http.StatusBadRequest)
+		return
+	}
+
+	ctx := appengine.NewContext(r)
+	token, err := config.Exchange(ctx, code)
+	if err != nil {
+		oauthCookieClear(w)
+		http.Redirect(w, r, cookieData.GenerateReturnURL(false), http.StatusFound)
+		return
+	}
+	if !token.Valid() {
+		oauthCookieClear(w)
+		http.Redirect(w, r, cookieData.GenerateReturnURL(false), http.StatusFound)
+		return
+	}
+
+	url := cookieData.GenerateReturnURL(true)
+	cookieData = &oauthCookie{
+		LoggedInWith: provider,
+		Token:        token,
+	}
+	cookie, err := cookieData.ToCookie()
+	if err != nil {
+		http.Error(w, "Failed to create cookie", http.StatusInternalServerError)
+		return
+	}
+
+	http.SetCookie(w, cookie)
+	http.Redirect(w, r, url, http.StatusFound)
+}
+
+func oauthGoogleCallback(w http.ResponseWriter, r *http.Request) {
+	ctx := appengine.NewContext(r)
+	config, err := getOauthGoogleConfig(ctx)
+	if err != nil {
+		http.Error(w, "Failed to get OAuth2 config", http.StatusInternalServerError)
+		return
+	}
+
+	oauthCallbackForConfig(w, r, config, "google")
+}
+
+func oauthStatus(w http.ResponseWriter, r *http.Request) {
+	loggedIn := false
+	cookie, err := oauthCookieFromRequest(r)
+	if err == nil {
+		loggedIn = cookie.LoggedInWith != ""
+	}
+
+	SendJsonResponse(w, map[string]interface{}{
+		"loggedIn": loggedIn,
+	})
+}
+
+func oauthLogout(w http.ResponseWriter, r *http.Request) {
+	returnURL := r.URL.Query().Get("return")
+	if returnURL == "" {
+		http.Error(w, "Missing return URL", http.StatusBadRequest)
+		return
+	}
+	returnURL += "#success=true"
+
+	oauthCookieClear(w)
+	http.Redirect(w, r, returnURL, http.StatusFound)
+}
+
+type oauthCookie struct {
+	State        string
+	ReturnURL    string
+	LoggedInWith string
+	Token        *oauth2.Token
+}
+
+func (c *oauthCookie) GenerateReturnURL(success bool) string {
+	if !success {
+		return c.ReturnURL + "#success=false"
+	}
+	return c.ReturnURL + "#success=true"
+}
+
+func (c *oauthCookie) ToCookie() (*http.Cookie, error) {
+	dataJson, err := json.Marshal(c)
+	if err != nil {
+		return nil, err
+	}
+	return &http.Cookie{
+		Name:     OAUTH_COOKIE,
+		Value:    base64.RawStdEncoding.EncodeToString(dataJson),
+		Path:     OAUTH_BASE,
+		HttpOnly: true,
+	}, nil
+}
+
+func oauthCookieFromRequest(r *http.Request) (*oauthCookie, error) {
+	cookie, err := r.Cookie(OAUTH_COOKIE)
+	if err != nil {
+		return nil, err
+	}
+	data, err := base64.RawStdEncoding.DecodeString(cookie.Value)
+	if err != nil {
+		return nil, err
+	}
+
+	var out oauthCookie
+	if err := json.Unmarshal(data, &out); err != nil {
+		return nil, err
+	}
+	return &out, nil
+}
+
+func oauthCookieClear(w http.ResponseWriter) {
+	http.SetCookie(w, &http.Cookie{
+		Name:     OAUTH_COOKIE,
+		Path:     OAUTH_BASE,
+		HttpOnly: true,
+		MaxAge:   -1,
+	})
+}
+
+func getOauthGoogleConfig(ctx context.Context) (*oauth2.Config, error) {
```

Comment:
suggestion: rename to `fetchOauthGoogleConfig` to make it clear that this involves a database request.